### PR TITLE
feat(amf): Logging improvements at function level

### LINKED ISSUE #12277: [NGAP][AMF]: To support profiling the major functions
Hi Priya,

Please uniform all the relevant functions with the below OAILOG macros to support for profiling and aid in debugging the issues.

OAILOG_FUNC_IN
OAILOG_FUNC_RETURN

Basic idea is to enable basic profiling with these macros to identify and optimize the heavily functions.

Regards,
Kaleem

### PR DESCRIPTION
**Pull request by priya-wavelabs**
Signed-off-by: priya-wavelabs <priya.agrawal@wavelabs.ai>

## Summary
Fix:
  Modified the functions entry and exit points under OAI_LOG for tracing.

## Test Plan

- Verified with unit test.
image

- Verified basic sanity with ueransim. 
  pcap snap :
MicrosoftTeams-image

MME Log : 
[mme6.log](https://github.com/magma/magma/files/8450734/mme6.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by ganeshg87 — APPROVED** [review]
lgtm..
please address CI misspell issue
https://github.com/magma/magma/runs/6002744273?check_suite_focus=true

File: lte/gateway/c/core/oai/tasks/amf/Registration.cpp [hunk 1]
```diff
@@ -691,6 +695,7 @@ int amf_send_registration_accept(amf_context_t* amf_context) {
 
 static int registration_accept_t3550_handler(zloop_t* loop, int timer_id,
                                              void* arg) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: At line, https://github.com/magma/magma/pull/12348/files#diff-5971eeebd29f80bdbd0931787b020513ed6287d30b033723dbff1f8c65e4de37R706, module name LOG_NAS_AMF is used.
Could you include the relevant module.

Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: Correct in other logging APIs also

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 2]
```diff
@@ -134,23 +145,28 @@ void AmfNasStateManager::create_state() {
...
   amf_nas_state_init_local_state();
+  OAILOG_FUNC_OUT(LOG_AMF_APP);
 }
 
 // Free the memory allocated to state pointer
 void AmfNasStateManager::free_state() {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
   if (!state_cache_p) {
-    return;
+    OAILOG_FUNC_OUT(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:156]: Add log to indicate that state_cache_p is NULL

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 3]
```diff
@@ -176,21 +193,24 @@ map_uint64_ue_context_t* AmfNasStateManager::get_ue_state_map() {
...
 void AmfNasStateManager::clear_db_state() {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
   OAILOG_DEBUG(LOG_AMF_APP, "Clearing state in data store");
   std::vector<std::string> keys_to_del;
   keys_to_del.emplace_back(AMF_NAS_STATE_KEY);
 
   if (redis_client->clear_keys(keys_to_del) != RETURNok) {
     OAILOG_ERROR(LOG_AMF_APP, "Failed to clear the state in data store");
-    return;
+    OAILOG_FUNC_OUT(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Return is not required

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 4]
```diff
@@ -176,21 +193,24 @@ map_uint64_ue_context_t* AmfNasStateManager::get_ue_state_map() {
...
+  OAILOG_FUNC_OUT(LOG_AMF_APP);
 }
 
 void put_amf_ue_state(amf_app_desc_t* amf_app_desc_p, imsi64_t imsi64,
                       bool force_ue_write) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
   if ((!AmfNasStateManager::getInstance().is_persist_state_enabled()) ||
       (imsi64 == INVALID_IMSI64)) {
-    return;
+    OAILOG_FUNC_OUT(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:213]: Add error log with message

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 5]
```diff
@@ -235,19 +257,21 @@ void AmfNasStateManager::write_ue_state_to_db(
...
   if (!persist_state_enabled) {
-    return RETURNok;
+    OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNok);
   }
   auto keys = redis_client->get_keys("IMSI*" + task_name + "*");
   for (const auto& key : keys) {
     magma::lte::oai::UeContext ue_proto = magma::lte::oai::UeContext();
     if (redis_client->read_proto(key.c_str(), ue_proto) != RETURNok) {
-      return RETURNerror;
+      OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:276]: Add error log with message

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 6]
```diff
@@ -257,29 +281,30 @@ status_code_e AmfNasStateManager::read_ue_state_from_db() {
...
   if (!persist_state_enabled) {
-    return RETURNok;
+    OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNok);
   }
   for (const auto& kv :
        AMFClientServicer::getInstance().map_imsi_ue_proto_str.umap) {
     magma::lte::oai::UeContext ue_proto = magma::lte::oai::UeContext();
     if (!ue_proto.ParseFromString(kv.second)) {
-      return RETURNerror;
+      OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:298]: Same as above

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 7]
```diff
@@ -293,19 +318,20 @@ status_code_e AmfNasStateManager::read_state_from_db() {
     if (AMFClientServicer::getInstance().map_table_key_proto_str.get(
             table_key, &proto_str) != magma::MAP_OK) {
       OAILOG_DEBUG(LOG_MME_APP, "Failed to read proto from db \n");
-      return RETURNerror;
+      OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
     }
     // Deserialization Step
     if (!state_proto.ParseFromString(proto_str)) {
-      return RETURNerror;
+      OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:329]: Add error log with message

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 8]
```diff
@@ -134,23 +145,29 @@ void AmfNasStateManager::create_state() {
...
   amf_nas_state_init_local_state();
+  OAILOG_FUNC_OUT(LOG_AMF_APP);
 }
 
 // Free the memory allocated to state pointer
 void AmfNasStateManager::free_state() {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
   if (!state_cache_p) {
-    return;
+    OAILOG_DEBUG(LOG_AMF_APP, "state_cache_p is NULL");
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Log with error

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 9]
```diff
@@ -496,12 +513,12 @@ void amf_app_state_free_ue_context(void** ue_context_node) {
  ***************************************************************************/
 tmsi_t amf_lookup_guti_by_ueid(amf_ue_ngap_id_t ue_id) {
   amf_context_t* amf_ctxt = amf_context_get(ue_id);
-
+  OAILOG_FUNC_IN(LOG_AMF_APP);
   if (amf_ctxt == NULL) {
-    return (0);
+    OAILOG_FUNC_RETURN(LOG_AMF_APP, 0);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:518]: I think we need to return "INVALID_TMSI"

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 10]
```diff
@@ -1007,7 +1009,7 @@ static int amf_as_security_req(const amf_as_security_t* msg,
         } else {
           OAILOG_ERROR(LOG_AMF_APP, "UE not found : " AMF_UE_NGAP_ID_FMT "\n",
                        as_msg->ue_id);
-          return -2;
+          OAILOG_FUNC_RETURN(LOG_AMF_APP, -2);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: What's the reason for returning "-2" What action is taken with returning "-2".

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp [hunk 11]
```diff
@@ -50,20 +50,27 @@ amf_app_desc_t* get_amf_nas_state(bool read_from_redis) {
...
+  AmfNasStateManager::getInstance().free_state();
+  OAILOG_FUNC_OUT(LOG_AMF_APP);
+}
 
 map_uint64_ue_context_t* get_amf_ue_state() {
   return AmfNasStateManager::getInstance().get_ue_state_map();
 }
 
 void delete_amf_ue_state(imsi64_t imsi64) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:73]: Remove LOG_MME_APP module in all places for amf task

File: lte/gateway/c/core/oai/tasks/amf/amf_authentication.cpp [hunk 12]
```diff
@@ -914,6 +923,7 @@ int amf_send_authentication_request(amf_context_t* amf_ctx,
 // Fetch the serving network name
 static int calculate_amf_serving_network_name(amf_context_t* amf_ctx,
                                               unsigned char* snni) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_authentication.cpp:926]: While returning, right  module is not included

File: lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp [hunk 13]
```diff
@@ -133,8 +136,9 @@ bool AMFClientServicerBase::set_smf_notification(
 }
 
 AMFClientServicer& AMFClientServicer::getInstance() {
+  OAILOG_FUNC_IN(LOG_NGAP);
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp]: Is this right module name?

File: lte/gateway/c/core/oai/tasks/amf/Registration.cpp (hunk 1 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: Comment addressed.

Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 2 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:156]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 3 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 5 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:276]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 6 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:298]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 7 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:329]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 8 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 11 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:73]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_authentication.cpp (hunk 12 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_authentication.cpp:926]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp (hunk 13 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp]: Comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp (hunk 4 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:213]: It is not an error condition. The UE state is synced if imsi64 is valid.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 9 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp:518]: The caller of this function expects "0" if guti is not found.
The value of "INVALID_TMSI" is UINT32_MAX.

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 10 above)
Comment (priya-wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: Changed it to RETURNerror.
Comment addressed.

**Comment by sebathomas** [comment]
It looks like this PR contains a lot of changes in addition what you're trying to do. You should probably create a new branch with just your own commits, and open a new PR.

**Review by xbend — CHANGES_REQUESTED** [review]
This PR contains merge commits.
Please rebase your commits on top of current master.

**Review by Neudrino — CHANGES_REQUESTED** [review]
Rebase on master needed here.

**Comment by priya-wavelabs** [comment]
I did a mistake while pushing to this PR, so I am discarding this PR.
I have raised separate PR #12652.
All the review comments given by  @[rsarwad](https://github.com/rsarwad)  have been addressed.
I updated PR #12652.

**Pull request closed without merging**